File: Scripts/hana_transferable.py
#!/usr/bin/env python3
"""Which off-subtype HA/NA markers are actually untransferable, and which are
only labelled that way?

FLUMUT_KEEP_MISMATCHED_HA_NA drops every HA/NA marker when the reference's
subtype does not match FluMut's H5/N1 numbering. The case for that is real but
it is an average: FluMut's reference and this run's carry different wild-type
residues at most HA1 positions, so a marker phrased as a deviation from the H5
residue is being read against a protein that never had it.

Per MARKER the answer is already known. flumut_position_map.tsv records, for
every position it places, whether the two references carry the same residue. A
marker sitting on an identical residue is a deviation from the same starting
point, and its H5 provenance is a label rather than a confound.

Reads a finished run and writes one row per dropped HA/NA marker with that
verdict, so the flag stops standing in for a judgement it cannot express.

Rows are counted AFTER reference subtraction — markers_all.tsv minus
reference_markers.tsv. Reading markers_all.tsv alone inflates the answer with
markers this run's own reference carries, which appear in nearly every sample.

usage: hana_transferable.py <run_dir> [out.tsv]
"""
import collections
import os
import re
import sys

MARK = re.compile(r'^([A-Za-z0-9\-]+):([A-Z])(\d+)([A-Z])$')


def rows_of(path, col=1):
    out = []
    if not os.path.exists(path):
        return out
    with open(path) as fh:
        next(fh, None)
        for line in fh:
            f = line.rstrip('\n').split('\t')
            if len(f) > col:
                out.append(f[col])
    return out


def main():
    run = sys.argv[1]
    fm = os.path.join(run, 'variant_analysis', 'flumut')
    out = sys.argv[2] if len(sys.argv) > 2 else os.path.join(fm, 'hana_transferable.tsv')

    pos = {}
    pmap = os.path.join(fm, 'flumut_position_map.tsv')
    if not os.path.exists(pmap):
        sys.exit(f'no position map at {pmap} -- run FLUMUT_POSITION_MAP first')
    with open(pmap) as fh:
        next(fh)
        for line in fh:
            f = line.rstrip('\n').split('\t')
            if len(f) >= 8:
                pos.setdefault(f[0], {})[int(f[2])] = (f[3], f[6], f[7])

    refm = set(rows_of(os.path.join(fm, 'reference_markers.tsv')))
    kept = set(rows_of(os.path.join(fm, 'markers.tsv')))
    counts = collections.Counter(m for m in rows_of(os.path.join(fm, 'markers_all.tsv'))
                                 if m not in refm)

    recs, same = [], 0
    for marker, n in sorted(counts.items(), key=lambda x: -x[1]):
        for tok in (t.strip() for t in marker.split(',')):
            g = MARK.match(tok)
            if not g or g.group(1)[:2] not in ('HA', 'NA'):
                continue
            label, wt, p = g.group(1), g.group(2), int(g.group(3))
            rec = pos.get(label, {}).get(p)
            if rec is None:
                recs.append((marker, tok, label, p, '', '', 'unplaced', n,
                             'yes' if marker in kept else 'no', ''))
                continue
            fm_aa, ref_aa, ident = rec
            if ident == 'yes':
                same += 1
            # The marker's own wild-type should be FluMut's residue. When it is
            # not, the marker is phrased against something other than the
            # reference the map placed, and the verdict is worth less.
            note = '' if wt == fm_aa else f'marker wild-type {wt}'
            recs.append((marker, tok, label, p, fm_aa, ref_aa, ident, n,
                         'yes' if marker in kept else 'no', note))

    with open(out, 'w') as fh:
        fh.write('marker\tcomponent\tlabel\tflumut_pos\tflumut_aa\tref_aa\t'
                 'wt_identical\tsample_rows\tkept\tnote\n')
        for r in recs:
            fh.write('\t'.join(str(x) for x in r) + '\n')

    # One row per COMPONENT, so a compound marker appears twice. Sample-rows are
    # summed over distinct markers instead, or the compound ones count double.
    distinct = {r[0]: r[7] for r in recs}
    print(f'HA/NA after reference subtraction: {len(distinct)} distinct markers, '
          f'{sum(distinct.values())} sample-rows, {len(recs)} components')
    print(f'  components on an IDENTICAL wild-type residue (transferable): {same}')
    print(f'  components on a DIFFERENT wild-type residue                : {len(recs) - same}')
    print(f'written to {out}')


if __name__ == '__main__':
    main()

File: Scripts/hana_transferable.R
#!/usr/bin/env Rscript

#### hana_transferable.R
####
#### Identify which off-subtype HA/NA markers are untransferable and which are
#### only labelled that way.
####
#### FLUMUT_KEEP_MISMATCHED_HA_NA drops every HA/NA marker when the reference's
#### subtype does not match FluMut's H5/N1 numbering. The case for that is real
#### but it is an average: FluMut's reference and this run's carry different
#### wild-type residues at most HA1 positions, so a marker phrased as a deviation
#### from the H5 residue is read against a protein that never had it.
####
#### The answer is known for each marker. flumut_position_map.tsv records, for
#### every position it places, whether the two references carry the same residue.
#### A marker on an identical residue is a deviation from the same starting point,
#### so its H5 provenance is a label rather than a confound.
####
#### Read a finished run and write one row per dropped HA/NA marker with that
#### verdict.
####
#### Rows are counted AFTER reference subtraction -- markers_all.tsv minus
#### reference_markers.tsv. markers_all.tsv alone inflates the answer with markers
#### this run's own reference carries, which appear in nearly every sample.
####
#### usage: Rscript hana_transferable.R <run_dir> [out.tsv]

MARK = "^([A-Za-z0-9-]+):([A-Z])([0-9]+)([A-Z])$"

# Return one column (1-based) of a headed TSV, skipping the header, or an empty
# vector when the file is absent.
rows.of = function(path, col = 2L) {
  if (!file.exists(path)) return(character(0))
  lines = readLines(path)
  if (length(lines) <= 1L) return(character(0))
  vals = character(0)
  for (line in lines[-1]) {
    f = strsplit(line, "\t", fixed = TRUE)[[1]]
    if (length(f) >= col) vals = c(vals, f[col])
  }
  vals
}

args = commandArgs(trailingOnly = TRUE)
if (length(args) < 1L) {
  cat("Usage: Rscript hana_transferable.R <run_dir> [out.tsv]\n", file = stderr())
  quit(status = 1)
}
run = args[1]
fm = file.path(run, "variant_analysis", "flumut")
out = if (length(args) >= 2L) args[2] else file.path(fm, "hana_transferable.tsv")

# Position map: label -> flumut_pos -> (flumut_aa, ref_aa, identical).
pmap = file.path(fm, "flumut_position_map.tsv")
if (!file.exists(pmap)) {
  cat("no position map at ", pmap, " -- run FLUMUT_POSITION_MAP first\n",
      sep = "", file = stderr())
  quit(status = 1)
}
pos = list()
plines = readLines(pmap)
for (line in plines[-1]) {
  f = strsplit(line, "\t", fixed = TRUE)[[1]]
  if (length(f) < 8L) next
  label = f[1]
  if (is.null(pos[[label]])) pos[[label]] = list()
  pos[[label]][[f[3]]] = c(f[4], f[7], f[8])   # flumut_aa, ref_aa, identical
}

refm = unique(rows.of(file.path(fm, "reference_markers.tsv")))
kept = unique(rows.of(file.path(fm, "markers.tsv")))

# Count markers in markers_all.tsv that the reference does not carry, keeping
# first-appearance order for ties.
all.markers = rows.of(file.path(fm, "markers_all.tsv"))
all.markers = all.markers[!(all.markers %in% refm)]
uniq = unique(all.markers)
count.vec = as.integer(table(factor(all.markers, levels = uniq)))
ord = order(-count.vec)
markers.sorted = uniq[ord]
counts.sorted = count.vec[ord]

recs = list()
same = 0L
for (m in seq_along(markers.sorted)) {
  marker = markers.sorted[m]
  n = counts.sorted[m]
  for (tok in trimws(strsplit(marker, ",", fixed = TRUE)[[1]])) {
    g = regmatches(tok, regexec(MARK, tok))[[1]]
    if (length(g) == 0L) next
    label = g[2]; wt = g[3]; p = g[4]
    if (!substr(label, 1L, 2L) %in% c("HA", "NA")) next
    is.kept = if (marker %in% kept) "yes" else "no"
    rec = if (!is.null(pos[[label]])) pos[[label]][[p]] else NULL
    if (is.null(rec)) {
      recs[[length(recs) + 1L]] = list(marker, tok, label, p, "", "",
                                       "unplaced", n, is.kept, "")
      next
    }
    fm.aa = rec[1]; ref.aa = rec[2]; ident = rec[3]
    if (ident == "yes") same = same + 1L
    # The marker's wild-type should match FluMut's residue. When it does not, the
    # marker is phrased against a different reference residue, so the verdict is
    # less reliable.
    note = if (wt == fm.aa) "" else paste0("marker wild-type ", wt)
    recs[[length(recs) + 1L]] = list(marker, tok, label, p, fm.aa, ref.aa,
                                     ident, n, is.kept, note)
  }
}

header = paste("marker", "component", "label", "flumut_pos", "flumut_aa",
               "ref_aa", "wt_identical", "sample_rows", "kept", "note",
               sep = "\t")
body = vapply(recs, function(r) paste(unlist(r), collapse = "\t"), character(1))
writeLines(c(header, body), out)

# One row per component, so a compound marker appears twice. Sum sample rows over
# distinct markers to avoid double-counting compound markers.
distinct.markers = unique(vapply(recs, function(r) r[[1]], character(1)))
distinct.sum = sum(counts.sorted[match(distinct.markers, markers.sorted)])
cat("HA/NA after reference subtraction: ", length(distinct.markers),
    " distinct markers, ", distinct.sum, " sample-rows, ", length(recs),
    " components\n", sep = "")
cat("  components on an IDENTICAL wild-type residue (transferable): ", same, "\n",
    sep = "")
cat("  components on a DIFFERENT wild-type residue                : ",
    length(recs) - same, "\n", sep = "")
cat("written to ", out, "\n", sep = "")
